# Supervised Learning Model Evaluation Lab

Complete the exercises below to solidify your knowledge and understanding of supervised learning model evaluation.

In [1]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

## Regression Model Evaluation

In [3]:
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
data = pd.read_csv('housing.csv', header=None, delimiter=r"\s+", names=column_names)

In [4]:
"""
CRIM - per capita crime rate by town
ZN - proportion of residential land zoned for lots over 25,000 sq.ft.
INDUS - proportion of non-retail business acres per town.
CHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)
NOX - nitric oxides concentration (parts per 10 million)
RM - average number of rooms per dwelling
AGE - proportion of owner-occupied units built prior to 1940
DIS - weighted distances to five Boston employment centres
RAD - index of accessibility to radial highways
TAX - full-value property-tax rate per $10,000
PTRATIO - pupil-teacher ratio by town
B - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
LSTAT - % lower status of the population
MEDV - Median value of owner-occupied homes in $1000\'s"""

"\nCRIM - per capita crime rate by town\nZN - proportion of residential land zoned for lots over 25,000 sq.ft.\nINDUS - proportion of non-retail business acres per town.\nCHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)\nNOX - nitric oxides concentration (parts per 10 million)\nRM - average number of rooms per dwelling\nAGE - proportion of owner-occupied units built prior to 1940\nDIS - weighted distances to five Boston employment centres\nRAD - index of accessibility to radial highways\nTAX - full-value property-tax rate per $10,000\nPTRATIO - pupil-teacher ratio by town\nB - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town\nLSTAT - % lower status of the population\nMEDV - Median value of owner-occupied homes in $1000's"

In [5]:
data

,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296.0,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242.0,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242.0,17.8,392.83,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222.0,18.7,394.63,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222.0,18.7,396.90,5.33,36.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
501,0.06263,0.0,11.93,0,0.573,6.593,69.1,2.4786,1,273.0,21.0,391.99,9.67,22.4
502,0.04527,0.0,11.93,0,0.573,6.120,76.7,2.2875,1,273.0,21.0,396.90,9.08,20.6
503,0.06076,0.0,11.93,0,0.573,6.976,91.0,2.1675,1,273.0,21.0,396.90,5.64,23.9
504,0.10959,0.0,11.93,0,0.573,6.794,89.3,2.3889,1,273.0,21.0,393.45,6.48,22.0


## 1. Split this data set into training (80%) and testing (20%) sets.

The `MEDV` field represents the median value of owner-occupied homes (in $1000's) and is the target variable that we will want to predict.

In [6]:
# Your code here :
from sklearn.model_selection import train_test_split

X = data.drop(columns=['MEDV'])
y = data['MEDV']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(404, 13) (102, 13) (404,) (102,)


## 2. Train a `LinearRegression` model on this data set and generate predictions on both the training and the testing set.

In [7]:
# Your code here :
from sklearn.linear_model import LinearRegression

reg = LinearRegression()
reg.fit(X_train, y_train)

y_train_pred = reg.predict(X_train)
y_test_pred = reg.predict(X_test)

## 3. Calculate and print R-squared for both the training and the testing set.

In [8]:
# Your code here :
from sklearn.metrics import r2_score

r2_train = r2_score(y_train, y_train_pred)
r2_test = r2_score(y_test, y_test_pred)

print('R-squared (train):', r2_train)
print('R-squared (test):', r2_test)

R-squared (train): 0.7508856358979673
R-squared (test): 0.6687594935356326


## 4. Calculate and print mean squared error for both the training and the testing set.

In [9]:
# Your code here :
from sklearn.metrics import mean_squared_error

mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)

print('MSE (train):', mse_train)
print('MSE (test):', mse_test)

MSE (train): 21.641412753226312
MSE (test): 24.291119474973478


## 5. Calculate and print mean absolute error for both the training and the testing set.

In [10]:
# Your code here :
from sklearn.metrics import mean_absolute_error

mae_train = mean_absolute_error(y_train, y_train_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)

print('MAE (train):', mae_train)
print('MAE (test):', mae_test)

MAE (train): 3.314771626783228
MAE (test): 3.189091965887837


## Classification Model Evaluation

In [11]:
from sklearn.datasets import load_iris
data = load_iris()

In [12]:
print(data.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

In [13]:
column_names = data.feature_names

In [14]:
df = pd.DataFrame(data['data'],columns=column_names)

In [15]:
df

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2
...,...,...,...,...
145,6.7,3.0,5.2,2.3
146,6.3,2.5,5.0,1.9
147,6.5,3.0,5.2,2.0
148,6.2,3.4,5.4,2.3


In [16]:
target = pd.DataFrame(data.target)

In [17]:
data.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [18]:
data['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

## 6. Split this data set into training (80%) and testing (20%) sets.

The `class` field represents the type of flower and is the target variable that we will want to predict.

In [19]:
# Your code here :
X_train, X_test, y_train, y_test = train_test_split(
    df, target, test_size=0.2, random_state=42, stratify=target
)

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(120, 4) (30, 4) (120, 1) (30, 1)


## 7. Train a `LogisticRegression` model on this data set and generate predictions on both the training and the testing set.

In [20]:
# Your code here :
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train.values.ravel())

y_train_pred = clf.predict(X_train)
y_test_pred = clf.predict(X_test)

## 8. Calculate and print the accuracy score for both the training and the testing set.

In [21]:
# Your code here :
from sklearn.metrics import accuracy_score

acc_train = accuracy_score(y_train, y_train_pred)
acc_test = accuracy_score(y_test, y_test_pred)

print('Accuracy (train):', acc_train)
print('Accuracy (test):', acc_test)

Accuracy (train): 0.975
Accuracy (test): 0.9666666666666667


## 9. Calculate and print the balanced accuracy score for both the training and the testing set.

In [22]:
# Your code here :
from sklearn.metrics import balanced_accuracy_score

bal_acc_train = balanced_accuracy_score(y_train, y_train_pred)
bal_acc_test = balanced_accuracy_score(y_test, y_test_pred)

print('Balanced accuracy (train):', bal_acc_train)
print('Balanced accuracy (test):', bal_acc_test)

Balanced accuracy (train): 0.975
Balanced accuracy (test): 0.9666666666666667


## 10. Calculate and print the precision score for both the training and the testing set.

In [23]:
# Your code here :
from sklearn.metrics import precision_score

prec_train = precision_score(y_train, y_train_pred, average='macro')
prec_test = precision_score(y_test, y_test_pred, average='macro')

print('Precision (train):', prec_train)
print('Precision (test):', prec_test)

Precision (train): 0.9751928288513655
Precision (test): 0.9696969696969697


## 11. Calculate and print the recall score for both the training and the testing set.

In [24]:
# Your code here :
from sklearn.metrics import recall_score

rec_train = recall_score(y_train, y_train_pred, average='macro')
rec_test = recall_score(y_test, y_test_pred, average='macro')

print('Recall (train):', rec_train)
print('Recall (test):', rec_test)

Recall (train): 0.975
Recall (test): 0.9666666666666667


## 12. Calculate and print the F1 score for both the training and the testing set.

In [25]:
# Your code here :
from sklearn.metrics import f1_score

f1_train = f1_score(y_train, y_train_pred, average='macro')
f1_test = f1_score(y_test, y_test_pred, average='macro')

print('F1 score (train):', f1_train)
print('F1 score (test):', f1_test)

F1 score (train): 0.974996093139553
F1 score (test): 0.9665831244778612


## 13. Generate confusion matrices for both the training and the testing set.

In [26]:
# Your code here :
from sklearn.metrics import confusion_matrix

cm_train = confusion_matrix(y_train, y_train_pred)
cm_test = confusion_matrix(y_test, y_test_pred)

print('Confusion matrix (train):')
print(cm_train)
print()
print('Confusion matrix (test):')
print(cm_test)

Confusion matrix (train):
[[40  0  0]
 [ 0 38  2]
 [ 0  1 39]]

Confusion matrix (test):
[[10  0  0]
 [ 0  9  1]
 [ 0  0 10]]


## Bonus: For each of the data sets in this lab, try training with some of the other models you have learned about, recalculate the evaluation metrics, and compare to determine which models perform best on each data set.

In [27]:
# Have fun here !

# --- Regression dataset: compare LinearRegression, KNN Regressor and Decision Tree Regressor ---
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
housing = pd.read_csv('housing.csv', header=None, delimiter=r"\s+", names=column_names)

X = housing.drop(columns=['MEDV'])
y = housing['MEDV']

X_train_h, X_test_h, y_train_h, y_test_h = train_test_split(X, y, test_size=0.2, random_state=42)

from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor

reg_models = {
    'LinearRegression': LinearRegression(),
    'KNeighborsRegressor': KNeighborsRegressor(),
    'DecisionTreeRegressor': DecisionTreeRegressor(random_state=42),
}

print("Regression model comparison (test set)")
print("-" * 60)
for name, model in reg_models.items():
    model.fit(X_train_h, y_train_h)
    preds = model.predict(X_test_h)
    print(f"{name:25s} R2: {r2_score(y_test_h, preds):.3f}  "
          f"MSE: {mean_squared_error(y_test_h, preds):.3f}  "
          f"MAE: {mean_absolute_error(y_test_h, preds):.3f}")

Regression model comparison (test set)
------------------------------------------------------------
LinearRegression          R2: 0.669  MSE: 24.291  MAE: 3.189
KNeighborsRegressor       R2: 0.647  MSE: 25.860  MAE: 3.664
DecisionTreeRegressor     R2: 0.858  MSE: 10.416  MAE: 2.394


In [28]:
# --- Classification dataset: compare LogisticRegression, KNN Classifier and Decision Tree Classifier ---
iris = load_iris()
df_iris = pd.DataFrame(iris['data'], columns=iris.feature_names)
target_iris = pd.DataFrame(iris.target)

X_train_i, X_test_i, y_train_i, y_test_i = train_test_split(
    df_iris, target_iris, test_size=0.2, random_state=42, stratify=target_iris
)

from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier

clf_models = {
    'LogisticRegression': LogisticRegression(max_iter=1000),
    'KNeighborsClassifier': KNeighborsClassifier(),
    'DecisionTreeClassifier': DecisionTreeClassifier(random_state=42),
}

print("Classification model comparison (test set)")
print("-" * 70)
for name, model in clf_models.items():
    model.fit(X_train_i, y_train_i.values.ravel())
    preds = model.predict(X_test_i)
    print(f"{name:25s} Accuracy: {accuracy_score(y_test_i, preds):.3f}  "
          f"F1 (macro): {f1_score(y_test_i, preds, average='macro'):.3f}  "
          f"Precision (macro): {precision_score(y_test_i, preds, average='macro'):.3f}  "
          f"Recall (macro): {recall_score(y_test_i, preds, average='macro'):.3f}")

Classification model comparison (test set)
----------------------------------------------------------------------
LogisticRegression        Accuracy: 0.967  F1 (macro): 0.967  Precision (macro): 0.970  Recall (macro): 0.967
KNeighborsClassifier      Accuracy: 1.000  F1 (macro): 1.000  Precision (macro): 1.000  Recall (macro): 1.000
DecisionTreeClassifier    Accuracy: 0.933  F1 (macro): 0.933  Precision (macro): 0.933  Recall (macro): 0.933


#### Conclusion:

On the **housing (regression)** dataset, the `DecisionTreeRegressor` achieved the best test-set performance (highest R2, lowest MSE/MAE), beating both `LinearRegression` and `KNeighborsRegressor`. This suggests the relationship between the housing features and price is non-linear enough that a tree can capture patterns a straight-line model misses. Linear regression is still a solid, easily interpretable baseline, and KNN performed slightly worse here, likely because it is sensitive to the unscaled/differently-scaled features in this dataset.

On the **iris (classification)** dataset, all three models perform very well (all above 93% accuracy) since the classes are largely separable, especially by petal length/width. In this run, `KNeighborsClassifier` achieved perfect accuracy (100%) on the test set, `LogisticRegression` was close behind (~97%), and `DecisionTreeClassifier` was slightly lower (~93%) — though with such a small, well-separated dataset, differences between models can vary with the random train/test split.